api/app/signals/apps/email_integrations/reporter/mail_actions.py

```diff
@@ -0,0 +1,214 @@
+                'reporter__email__isnull': False,
+                'reporter__email__gt': 0
+            },
+        },
+        'kwargs': {
+            'subject': 'Betreft melding: {signal_id}',
+            'templates': {
+                'txt': 'email/signal_status_changed_ingepland.txt',
+                'html': 'email/signal_status_changed_ingepland.html'
+            }
+        }
+    },
+    {
+        'name': 'Sent mail signal reopened',
+        'conditions': {
+            'filters': {
+                'status__state__in': [HEROPEND, ],
+                'reporter__email__isnull': False,
+                'reporter__email__gt': 0
+            },
+        },
+        'kwargs': {
+            'subject': 'Betreft melding: {signal_id}',
+            'templates': {
+                'txt': 'email/signal_status_changed_heropend.txt',
+                'html': 'email/signal_status_changed_heropend.html'
+            }
+        }
+    }
+]
+
+
+def _create_feedback_and_mail_context(signal: Signal):
+    """
+    Util functions to create the feedback object and create the context needed for the mail
+
+    :param signal:
+    :return:
+    """
+    from signals.apps.feedback.models import Feedback
+    from signals.apps.feedback.utils import get_feedback_urls
+
+    feedback = Feedback.actions.request_feedback(signal)
+    positive_feedback_url, negative_feedback_url = get_feedback_urls(feedback)
+    return {
+        'negative_feedback_url': negative_feedback_url,
+        'positive_feedback_url': positive_feedback_url,
+    }
+
+
+class MailActions:
+    _from_email = settings.NOREPLY
+
+    def __init__(self) -> None:
+        self._actions = {}
+        self._conditions = {}
+        self._kwargs = {}
+
+        for config in MAIL_CONFIGURATION:
+            key = slugify(config['name'])
+
+            self._actions[key] = config['name']
+            self._conditions[key] = config['conditions']
+            self._kwargs[key] = config['kwargs'] or {}
+
+    def _apply_filters(self, filters: dict, signal: Signal) -> bool:
+        try:
+            Signal.objects.filter(**filters).get(pk=signal.pk)
+            return True
+        except Signal.DoesNotExist:
+            pass
+        return False
+
+    def _apply_functions(self, functions: dict, signal: Signal) -> bool:
+        return all([
+            function(signal)
+            for _, function in functions.items()
+        ])
+
+    def _apply_conditions(self, conditions: dict, signal: Signal) -> Any:
+        filters = conditions['filters'] if 'filters' in conditions else {}
+        functions = conditions['functions'] if 'functions' in conditions else {}
+
+        return (self._apply_filters(filters=filters, signal=signal) and
+                self._apply_functions(functions=functions, signal=signal))
+
+    def _get_actions(self, signal: Signal) -> list:
+        found_actions_to_apply = []
+        for action, name in self._actions.items():
```

Comment:
This does not read nicely, `self._actions` contains a mapping from a key/code to the name of an action name that is then not used. Get rid of the `.items()` call `.keys()` or just iterate over the dictionary directly to get only the keys.

Or even, since this code assumes there is a conditions entry for each mail config just loop over `self._conditions.items()` and keep track of the keys while applying the conditions.

Reply:
Refactored